File: tests/Bus/ChangePrices.ts
import * as anchor from "@project-serum/anchor";
import { BusTripTicket } from "../../target/types/bus_trip_ticket";
import { BusLine } from "../Accounts"

describe("Take a Trip", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const program = anchor.workspace.BusTripTicket as anchor.Program<BusTripTicket>

  it("Buying Ticket", async () => {
    const Account = await program.account.busAccount.fetch(BusLine);
    const tx = await program.methods.changePrices(
      new anchor.BN(1785714),
      new anchor.BN(2000000),
      new anchor.BN(2100000),
      new anchor.BN(2200000),
      new anchor.BN(2300000)
    )
    .accounts({
      bus: BusLine,
      signer: provider.publicKey,
  })
    .rpc();
    console.log("----------------------------------------------")
    console.log("Your transaction signature", tx);
    console.log("----------------------------------------------")
    console.log("PDA: ", BusLine.toBase58())
    console.log("----------------------------------------------")
    console.log("Authority:", Account.authority.toBase58())
    console.log("----------------------------------------------")
    console.log("Bump:", Account.bumpOriginal.toString())
    console.log("----------------------------------------------")
    console.log("0 to 3 km ticket price:", Account.toThreeKm.toString())
    console.log("----------------------------------------------")
    console.log("3 to 6 km ticket price:", Account.toSixKm.toString())
    console.log("----------------------------------------------")
    console.log("6 to 12 km ticket price:", Account.toTwelveKm.toString())
    console.log("----------------------------------------------")
    console.log("12 to 27 km ticket price:", Account.toTwentySevenKm.toString())
    console.log("----------------------------------------------")
    console.log("more 27 km ticket price:", Account.moreTwentySevenKm.toString())
    console.log("----------------------------------------------")
  });
});

File: programs/sube-crypto/src/instructions/change_prices.rs
use anchor_lang::prelude::*;
use crate::errors::ErrorCode; 
use crate::state::accounts::*;

pub fn change_prices(
    ctx: Context<ChangePrices>,
    to_three_km: u64,
    to_six_km: u64,
    to_twelve_km: u64,
    to_twenty_seven_km: u64,
    more_twenty_seven_km: u64
) -> Result<()> {
    require_keys_eq!(ctx.accounts.signer.key(), ctx.accounts.bus.authority.key(), ErrorCode::AuthorityError);
    let bus: &mut Account<BusAccount> = &mut ctx.accounts.bus;
    bus.to_three_km = to_three_km;
    bus.to_six_km = to_six_km;
    bus.to_twelve_km = to_twelve_km;
    bus.to_twenty_seven_km = to_twenty_seven_km;
    bus.more_twenty_seven_km = more_twenty_seven_km;
    Ok(())
}

#[derive(Accounts)]
pub struct ChangePrices<'info> {
    #[account(mut, seeds = [bus.authority.key().as_ref()], bump = bus.bump_original)]
    pub bus: Account<'info, BusAccount>,
    #[account(mut)]
    pub signer: Signer<'info>,
}